lib/notifier/index.js

```diff
@@ -0,0 +1,384 @@
+    const zeroConfLimit = cashOutConfig.zeroConfLimit
+    const zeroConf = isCashOut && tx.fiat.lte(zeroConfLimit)
+    const notificationsEnabled = notifSettings.sms.transactions || notifSettings.email.transactions
+    const customerPromise = tx.customerId ? customers.getById(tx.customerId) : Promise.resolve({})
+
+    if (!notificationsEnabled && !highValueTx) return Promise.resolve()
+    if (zeroConf && isCashOut && !rec.isRedemption && !rec.error) return Promise.resolve()
+    if (!zeroConf && rec.isRedemption) return sendRedemptionMessage(tx.id, rec.error)
+
+    return Promise.all([
+      machineLoader.getMachineName(tx.deviceId),
+      customerPromise
+    ]).then(([machineName, customer]) => {
+      return utils.buildTransactionMessage(tx, rec, highValueTx, machineName, customer)
+    }).then(([msg, highValueTx]) => sendTransactionMessage(msg, highValueTx))
+  })
+}
+
+function sendRedemptionMessage (txId, error) {
+  const subject = `Here's an update on transaction ${txId}`
+  const body = error
+    ? `Error: ${error}`
+    : 'It was just dispensed successfully'
+
+  const rec = {
+    sms: {
+      body: `${subject} - ${body}`
+    },
+    email: {
+      subject,
+      body
+    }
+  }
+  return sendTransactionMessage(rec)
+}
+
+function sendTransactionMessage (rec, isHighValueTx) {
+  return settingsLoader.loadLatest().then(settings => {
+    const notifications = configManager.getGlobalNotifications(settings.config)
+
+    const promises = []
+
+    const emailActive =
+      notifications.email.active &&
+      (notifications.email.transactions || isHighValueTx)
+    if (emailActive) promises.push(emailFuncs.sendMessage(settings, rec))
+
+    const smsActive =
+      notifications.sms.active &&
+      (notifications.sms.transactions || isHighValueTx)
+    if (smsActive) promises.push(smsFuncs.sendMessage(settings, rec))
+
+    return Promise.all(promises)
+  })
+}
+
+const clearOldCryptoNotifications = (balances) => {
+  return queries.getAllValidNotifications(CRYPTO_BALANCE).then(res => {
+    const notifications = _.map(it => {
+      return {
+        id: it.id,
+        cryptoCode: it.detail.cryptoCode,
+        code: it.detail.code
+      }
+    }, res)
+    const duplicateIndexes = []
+    const idsToInvalidate = []
+    _.forEach(notification => {
+      const idx = _.findIndex(balance => {
+        return balance.code === notification.code && balance.cryptoCode === notification.cryptoCode
+      }, balances)
+
+      if (idx === -1) {
+        // if notification in DB doesnt exist in balances anymore then it is invalid now
+        idsToInvalidate.push(notification.id)
+      } else {
+        // if it exists then it is a duplicate, add it to array
+        duplicateIndexes.push(idx)
+      }
+    }, notifications)
+    return (idsToInvalidate.length > 0 ? queries.batchInvalidate(idsToInvalidate) : Promise.resolve()).then(() => duplicateIndexes)
```

Comment:
This is kinda of a mess. If there's IDs to invalidate we don't return duplicateIndexes, is there a reason for that?

Reply:
it's to prevent readding an existing notification, that wasn't invalidated. It could use a better name and implementation anyway, so I fixed it